src/System.Net.Http/src/System/Net/Http/Managed/ConnectHelper.cs

```diff
@@ -105,16 +105,31 @@ public BuilderAndCancellationTokenSocketAsyncEventArgs(CancellationToken cancell
 
         public static async ValueTask<SslStream> EstablishSslConnectionAsync(HttpConnectionSettings settings, string host, HttpRequestMessage request, Stream stream, CancellationToken cancellationToken)
         {
-            RemoteCertificateValidationCallback callback = null;
-            if (settings._serverCertificateCustomValidationCallback != null)
+            // Create the options bag to use.  As we mutate it, we don't just use the shared instance.
+            SslClientAuthenticationOptions sslOptions = settings._sslOptions != null ?
+                settings._sslOptions.ShallowClone() :
+                new SslClientAuthenticationOptions();
+
+            // Use the specified host, unless the options explicitly overrides it.
+            if (sslOptions.TargetHost == null)
             {
-                callback = (object sender, X509Certificate certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors) =>
+                sslOptions.TargetHost = host;
+            }
+
+            // If there's a cert validation callback, wrap the original delegate in order to:
+            // 1) Change the sender to be the request message (expected by HttpClientHandler's delegate), and
+            // 2) Wrap any exceptions in an HttpRequestException.
+            RemoteCertificateValidationCallback callback = sslOptions.RemoteCertificateValidationCallback;
+            if (callback != null)
+            {
+                // Replace the delegate, which will call to the old one.
+                sslOptions.RemoteCertificateValidationCallback = (object sender, X509Certificate certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors) =>
                 {
                     try
                     {
-                        return settings._serverCertificateCustomValidationCallback(request, certificate as X509Certificate2, chain, sslPolicyErrors);
+                        return callback(request, certificate as X509Certificate2, chain, sslPolicyErrors);
                     }
-                    catch (Exception e)
+                    catch (Exception e) when (!(e is HttpRequestException))
```

Comment:
Why filter out the HttpRequestExceptions?

Reply:
My understanding of this catch block is to ensure that everything that escapes is an HttpRequestException; if we already have an HttpRequestException, we don't need to wrap it in one.  Or you think we should just always wrap and in such (rare) situations end up with an HttpRequestException that wraps an HttpRequestException?  I'm fine either way.

Reply:
Yeah, the latter.  I'm trying to imagine how you'd have an HttpRequestException here and not coming up with anything.  So if that ever does happen, seems fine to wrap it in another HttpRequestException.

Reply:
>So if that ever does happen, seems fine to wrap it in another HttpRequestException.

If we can repro a situation like this, we should fix it as a bug to that there is a single HttpRequestException as top-level and a more appropriate exception inside that.

Reply:
Actually, I'm not sure why wrapping the exception here is necessary at all.  Presumably, any exception here will end up bubbling out of AuthenticateAsClientAsync in some form (or not? could be that these are eaten, I don't know).  We should already be propagating exceptions from AuthenticateAsClientAsync appropriately (e.g. wrapping in HttpRequestException or whatever).  So why do we need to do anything with the exception here?